### ISSUE TITLE
I2C issue where two consecutive start bits are generated when reading in memory mode

### ISSUE THREAD

**Issue by zirus2402** [STMicroelectronics/STM32CubeF4#61]
**Setup**
 * STM32F401RE
 * TrueSTUDIO
 * STM32CubeF4 v1.25.2

**Bug**
In I2C memory mode, after the memory address has been written, the device should resend one start bit before switching to read mode. However, calling `HAL_I2C_Mem_Read_IT()` generates multiple start bits, resulting in the slave device answering with a NAK (see screenshot below). The event handler `HAL_I2C_EV_IRQHandler()` seems to be called too quickly, and the code handling the start bit is executed twice.

**Fix**
Incrementing `EventCount` in `I2C_MemoryTransmit_TXE_BTF()` (`stm32f4xx_hal_i2c.c` line 5282) prevents the start bit to be generated twice:

```C
...

else if (hi2c->EventCount == 2U)
  {
    if (CurrentState == HAL_I2C_STATE_BUSY_RX)
    {
      /* Generate Restart */
      hi2c->Instance->CR1 |= I2C_CR1_START;
      hi2c->EventCount++;
    }
    else if ((hi2c->XferCount > 0U) && (CurrentState == HAL_I2C_STATE_BUSY_TX))

    ...
```

**Screenshots**
Before the fix:
![two_start_bits](https://user-images.githubusercontent.com/15196917/108262065-7169e180-7132-11eb-9d62-2ebcc7849e1c.png)

After the fix:
![normal_behavior](https://user-images.githubusercontent.com/15196917/108262109-7b8be000-7132-11eb-8a96-9b82748930fe.png)

**Comment by ASELSTM** [comment]
Hi @zirus2402,

Thank you for this report and for the proposed fix. The issue you pointed out has been confirmed, a fix will be implemented and made available in future release. Thank you once again for your contribution.

With regards,

**Comment by ASELSTM** [comment]
ST Internal Reference: 102310

**Comment by ASELSTM** [comment]
Hi @zirus2402,

The issue you reported has been fixed in the frame of version v1.26.2 of the STM32CubeF4 published on GitHub.

Please allow me then to close this thread and thank you again for having reported.

With regards,
